awscli/customizations/history/list.py

```diff
@@ -0,0 +1,413 @@
+        'Return Code': StringTableCellFormatter,
+    }
+
+    def create(self, col, value, space):
+        cls = self._COL_TO_FORMATTER_TYPE[col]
+        packer = cls(value, space)
+        return packer
+
+
+class TableFormatter(object):
+    _MAXIMUM_FIELD_WIDTH = 50
+    _COMPONENT_COLORS = {
+        'delim': colorama.Style.BRIGHT + colorama.Fore.BLUE
+    }
+
+    def __init__(self, colorize=True):
+        self._factory = TableCellFormatterFactory()
+        self._colorize = colorize
+        if self._colorize:
+            colorama.init(autoreset=True, strip=False)
+
+    def __call__(self, title, records, stream):
+        if records:
+            self._draw_table(title, records, stream)
+
+    def _get_color(self, key):
+        return self._COMPONENT_COLORS.get(key)
+
+    def _draw_table(self, title, records, stream):
+        column_widths = self._get_column_widths(records)
+        # The number of columns + 1 is the number of separators we need
+        total_width = sum(column_widths.values()) + len(column_widths) + 1
+        self._write_title(stream, title, column_widths, total_width)
+        self._write_column_headers(stream, column_widths)
+        self._write_data_rows(stream, records, column_widths)
+
+    def _get_column_widths(self, records):
+        longest_value = OrderedDict()
+        record = records[0]
+        for header in record.keys():
+            longest_value[header] = len(header)
+        for record in records:
+            for header, value in record.items():
+                prior_longest = longest_value[header]
+                if isinstance(value, list):
+                    length = len(' '.join(value))
+                else:
+                    length = len(value)
+                longest_value[header] = min(max(prior_longest, length),
+                                            self._MAXIMUM_FIELD_WIDTH)
+
+        # Add room in each column for a space before and after the value
+        for key, value in longest_value.items():
+            longest_value[key] = value + 2
+        return longest_value
+
+    def _align_center(self, value, field_width):
+        padding_left = int(field_width / 2) - int(len(value) / 2)
+        padding_right = field_width - padding_left - len(value)
+        prefix = ' ' * padding_left
+        postfix = ' ' * padding_right
+        aligned_value = prefix + value + postfix
+        return aligned_value
+
+    def _align_left(self, value, field_width, offset=1):
+        prefix = ' ' * offset
+        padding_right = field_width - len(value) - offset
+        postfix = ' ' * padding_right
+        aligned_value = prefix + value + postfix
+        return aligned_value
+
+    def _write_title(self, stream, title, column_widths, total_width):
+        title_space = total_width - 2  # Leave space for the sides
+        centered_title = self._align_center(title, title_space)
+        self._write(stream, '-' * total_width + '\n', 'delim')
```

Comment:
Hmmm I think there is another layer of abstraction missing. I think it would make sense if we had two abstractions here: the `TableFormatter` that knows what is going to write and a `TableStyler` which is a helper class that the `TableFormatter` will call out to when writing. Having the the `TableFormatter` do the writes as well makes it a little difficult to read especially when the `write()` takes in a type as well.


I could imaging this `TableStyler` class look such as the following:
```python
class TableStyler(object):
      def __init__(self, ouput_stream, colorize=True):

      def column(self):
      # Writes the delimeter

     def row(self, length):
     # Writes the - for the row

    def align_cell_left(self, value, length):
    def align_cell_center(self, value, length):
```
That way the `TableFormatter` does not need to worry about the nitty gritty parts of writing out lower level spacing and components of the table.